Settings menu MVP

Fruitsapje opened the pull request:
This is a MVP setup for the settings menu in `video.js(5.*)`.

In order to make this plugin work an event must be added (`this.trigger('menuitemclicked');`) in the MenuItem handleClick function. Without this event this plugin won't work. If desired a PR for this event can made by us.

During the development I have come across some things with the PlaybackRateMenu that might be worth mentioning. The `PlaybackRateMenu` isn't set as `selectable` and thus doesn't get a `vjs-selected` class. Therefore I have created an extra check in the `SettingsMenuItem update()`  to update the SubMenuValue when a playback rate option is selected.

The styling of the settings menu is open for discuss, for now I have created a styling that fits with the MVP. Note that the settings menu currently has a large width since the settings sub menus labels currently come from thecontroltexts of the menus. Beside that, we can also take a look at the way the sub menu opens up and if we want to change that around or keep it the way it currently is.

Furthermore the tests are currently empty since the settings menu is a MVP for now and prone to change. Once the settings menu is settled a bit, tests can be added.

Commit:
Run video.js plugin generator

Commit:
MVP settings menu

Commit:
npm lint fixes for vjs-settings-menu

Commit:
removed utils.js

Commit:
Add dutch language file

gkatsev commented:
Here's a screenshot of what's current available:
![](https://cldup.com/5ATJf0YUVY.png)

gkatsev commented on src/components/settings-menu-item.js:
```diff
@@ -0,0 +1,141 @@
+
+const MenuItem = videojs.getComponent('MenuItem');
+const playbackRateMenuButton = videojs.getComponent('PlaybackRateMenuButton');
+const settingsSubMenuTitleEl = Symbol('settingsSubMenuTitleEl');
```
Is `Symbol` going to be available via babel in IE8?
I think it's fine if it doesn't work there for new features but if we can support it with _minimal_ work, that would be good.

gkatsev commented on src/components/settings-menu-item.js:
```diff
@@ -0,0 +1,141 @@
+    };
+
+    for (let item of this.subMenu.menu.children()) {
+      item.on('menuitemclicked', updateAfterTimeout);
```
here we can probably just listen to `click` like a normal element.

gkatsev commented on src/components/settings-menu-button.js:
```diff
@@ -0,0 +1,80 @@
+
+        // Hide children to avoid sub menus stacking on top of each other
+        // or having multiple menus open
+        settingsMenuItem.on('menuitemclicked', videojs.bind(this, this.hideChildren));
```
What about here listening to `click` and in `hideChildren` only hide the submenu of the non-selected items? This would mean that we want to have a concept of "selected" settings menu item. And I think it makes sense to have because the selected one is the one that has its sub menus open.

nickygerritsen replied:
Note that we also use `hideChildren` on `mouseleave` of the settings menu and in that case we _should_ hide all submenu's. So then there should be some distinction between these two cases.

gkatsev commented:
While, I'm not sure we want to have a `menuitemclicked` since we already have a `click` event, it may be useful to add a `change` event to menus that are triggered when the selected item is changed. This could also simplify the logic for hiding the submenus and updating the settings menu items.

Commit:
Removed Symbols and change menuitemclicked event to click

Fruitsapje commented:
I have removed the usage of Symbols and changed the `menuitemclicked` event to `click`. In order to make the plugin work correctly with the `click` event, I also added a function to add an open class to the 'selected' submenu and updated the `hideSubMenu` function with this class.

Also I noticed the checks that have been added to this PR have failed. According to @nickygerritsen this is because the plugin has no tests.

Fruitsapje commented:
Hey @gkatsev I was wondering if you had any chance to look at this PR? Also from this friday on I wont be able to respond for 2 weeks, just so you know :)

gkatsev commented:
Code-wise looks good! Haven't had a chance to give it another spin, though.

Fruitsapje commented:
Any possible ETA when you can give this another spin @gkatsev ? We'd like to implement this plugin ourselves aswell :)

gkatsev commented:
Finally took a look, sorry for the wait. LGTM!

nickygerritsen commented:
:tada:
